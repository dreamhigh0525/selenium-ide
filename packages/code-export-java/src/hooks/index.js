import dependencies from './declareDependencies'
import variables from './declareVariables'
import beforeAll from './beforeAll'
import beforeEach from './beforeEach'
import afterEach from './afterEach'
import afterAll from './afterAll'
import inEach from './inEach'

export default {
  dependencies,
  variables,
  beforeAll,
  beforeEach,
  afterEach,
  afterAll,
  inEach,
}
