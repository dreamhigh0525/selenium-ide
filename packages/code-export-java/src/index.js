// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import exporter from 'code-export-utils'
import Command from './command'
import { sanitizeName, capitalize } from './parsers'
import Hook from './hook'

const fileExtension = '.java'
export const commandPrefixPadding = '  '
const hooks = Hook.generateAll()

export async function emitTest({ baseUrl, test, tests }) {
  global.baseUrl = baseUrl
  const result = await _emitTest(test, tests)
  return {
    filename: generateFilename(test.name),
    body: _emitClass(test.name, result),
  }
}

export async function emitSuite({ baseUrl, suite, tests }) {
  global.baseUrl = baseUrl
  let result = ''
  for (const testName of suite.tests) {
    const test = tests.find(test => test.name === testName)
    result += await _emitTest(test, tests)
  }
  return {
    filename: generateFilename(suite.name),
    body: _emitClass(suite.name, result),
  }
}

function generateFilename(name) {
  return `${capitalize(sanitizeName(name))}${fileExtension}`
}

async function registerReusedTestMethods(test, tests) {
  for (const command of test.commands) {
    if (command.command === 'run') {
      const reusedTest = tests.find(test => test.name === command.target)
      const commands = reusedTest.commands.map(command => {
        return Command.emit(command)
      })
      const emittedCommands = await Promise.all(commands)
      const methodDeclaration = `public void ${reusedTest.name}() {`
      if (!hooks.declareMethods.isRegistered(methodDeclaration)) {
        hooks.declareMethods.register(methodDeclaration)
        hooks.declareMethods.register(
          emittedCommands
            .join(`\n${commandPrefixPadding}`)
            .replace(/^/, commandPrefixPadding)
        )
        hooks.declareMethods.register(`}`)
      }
    }
  }
}

function prettifyString(input, level) {
  return commandPrefixPadding.repeat(level) + input
}

async function _emitTest(test, tests) {
  let result = ''
  result += prettifyString('@Test\n', 1)
  result += prettifyString(`public void ${sanitizeName(test.name)}() {\n`, 1)
  let startingLevel = 2
  result += _renderHook(hooks.inEachBegin.emit({ isOptional: true }), {
    startingLevel,
  })
  await registerReusedTestMethods(test, tests)
  const commands = test.commands.map(command => {
    return Command.emit(command)
  })
  const emittedCommands = await Promise.all(commands)
  emittedCommands.forEach(emittedCommand => {
    const commandBlock = exporter.prettify(emittedCommand, {
      commandPrefixPadding,
      startingLevel,
    })
    startingLevel = commandBlock.endingLevel
    result += commandBlock.body
    result += '\n'
  })
  result += _renderHook(hooks.inEachEnd.emit({ isOptional: true }), {
    startingLevel,
  })
  result += prettifyString(`}\n\n`, 1)
  return result
}

function _renderHook(hook, { startingLevel } = { startingLevel: 1 }) {
  const result = exporter.prettify(hook, {
    commandPrefixPadding,
    startingLevel,
  }).body
  return result.trim().length ? result + '\n\n' : ''
}

function _emitClass(name, body) {
  let result = ''
  result += _renderHook(hooks.declareDependencies.emit(), { startingLevel: 0 })
  result += `public class ${capitalize(sanitizeName(name))} {\n`
  result += _renderHook(hooks.declareVariables.emit())
  result += _renderHook(hooks.beforeAll.emit({ isOptional: true }))
  result += _renderHook(hooks.beforeEach.emit())
  result += _renderHook(hooks.afterEach.emit())
  result += _renderHook(hooks.afterAll.emit({ isOptional: true }))
  result += _renderHook(hooks.declareMethods.emit({ isOptional: true }), {
    startingLevel: 0,
  })
  result += body
  result += `}`
  Hook.clear(hooks)
  return result
}

export default {
  emit: {
    test: emitTest,
    suite: emitSuite,
  },
  register: {
    command: Command.register,
    variable: hooks.declareVariables.register,
    dependency: hooks.declareDependencies.register,
    beforeAll: hooks.beforeAll.register,
    beforeEach: hooks.beforeEach.register,
    afterEach: hooks.afterEach.register,
    afterAll: hooks.afterAll.register,
    inEachBegin: hooks.inEachBegin.register,
    inEachEnd: hooks.inEachEnd.register,
    methods: hooks.declareMethods.register,
  },
}
