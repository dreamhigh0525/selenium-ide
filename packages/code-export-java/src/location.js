import exporter from '../../code-export-utils/src'

const emitters = {
  id: emitId,
  name: emitName,
  link: emitLink,
  linkText: emitLink,
  partialLinkText: emitPartialLinkText,
  css: emitCss,
  xpath: emitXpath,
}

export function emit(location) {
  return exporter.emit.location(location, emitters)
}

export default {
  emit,
}

function emitId(selector) {
  return `By.id("${selector}")`
}

function emitName(selector) {
  return `By.name("${selector}")`
}

function emitLink(selector) {
  return `By.linkText("${selector}")`
}

function emitPartialLinkText(selector) {
  return `By.partialLinkText("${selector}")`
}

function emitCss(selector) {
  return `By.cssSelector("${selector}")`
}

function emitXpath(selector) {
  return `By.xpath("${selector}")`
}
