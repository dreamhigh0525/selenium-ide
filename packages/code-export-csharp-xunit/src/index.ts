// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import {
  codeExport as exporter,
  generateHooks,
  LanguageEmitter,
  LanguageEmitterOpts,
} from '@seleniumhq/side-code-export'
import emitter from './command'
import hooks from './hook'
import { location } from '@seleniumhq/code-export-csharp-commons'

// Define language options
export const displayName = 'C# xUnit'

export const opts: LanguageEmitterOpts = {
  emitter: emitter,
  displayName,
  name: 'csharp-xunit',
  hooks: generateHooks(hooks),
  fileExtension: '.cs',
  commandPrefixPadding: '  ',
  terminatingKeyword: '}',
  commentPrefix: '//',
  generateFilename: (name) => {
    return `${exporter.parsers.capitalize(
      exporter.parsers.sanitizeName(name)
    )}Test${opts.fileExtension}`
  },

  generateMethodDeclaration: (name) => {
    return `public void ${exporter.parsers.sanitizeName(name)}() \n{`
  },
  generateTestDeclaration: (name) => {
    return `[Fact]\npublic void ${exporter.parsers.capitalize(
      exporter.parsers.sanitizeName(name)
    )}() {`
  },
  generateSuiteDeclaration: () => {
    return `public class SuiteTests : IDisposable {`
  },
}

const language: LanguageEmitter = {
  emit: {
    test: async function emitTest({
      test,
      tests,
      project,
      enableOriginTracing,
      beforeEachOptions,
      enableDescriptionAsComment,
    }) {
      const testDeclaration = opts.generateTestDeclaration(test.name)
      const result = await exporter.emit.test(test, tests, {
        ...opts,
        testDeclaration,
        enableOriginTracing,
        enableDescriptionAsComment,
        project,
      })
      const suiteName = test.name
      const suiteDeclaration = opts.generateSuiteDeclaration(suiteName)
      var _suite = await exporter.emit.suite(result, tests, {
        ...opts,
        suiteDeclaration,
        suiteName,
        project,
        beforeEachOptions,
      })
      return {
        filename: opts.generateFilename(test.name),
        body: exporter.emit.orderedSuite(_suite),
      }
    },
    suite: async function emitSuite({
      suite,
      tests,
      project,
      enableOriginTracing,
      beforeEachOptions,
      enableDescriptionAsComment,
    }) {
      const result = await exporter.emit.testsFromSuite(tests, suite, opts, {
        enableOriginTracing,
        enableDescriptionAsComment,
        generateTestDeclaration: opts.generateTestDeclaration,
        project,
      })
      const suiteDeclaration = opts.generateSuiteDeclaration(suite.name)
      var _suite = await exporter.emit.suite(result, tests, {
        ...opts,
        suiteDeclaration,
        suite,
        suiteName: suite.name,
        project,
        beforeEachOptions,
      })
      return {
        filename: opts.generateFilename(suite.name),
        body: emitOrderedSuite(_suite, suite.name),
      }
    },
    locator: location.emit,
    command: emitter.emit,
  },
  opts,
  register: {
    command: emitter.register,
    variable: opts.hooks.declareVariables.register,
    dependency: opts.hooks.declareDependencies.register,
    beforeAll: opts.hooks.beforeAll.register,
    beforeEach: opts.hooks.beforeEach.register,
    afterEach: opts.hooks.afterEach.register,
    afterAll: opts.hooks.afterAll.register,
    inEachBegin: opts.hooks.inEachBegin.register,
    inEachEnd: opts.hooks.inEachEnd.register,
  },
}

function generateNamespace(name: string) {
  return `namespace ${exporter.parsers.capitalize(
    exporter.parsers.sanitizeName(name)
  )}\n{\n`
}

function emitOrderedSuite(emittedSuite: any, name: string) {
  let result = ''
  result += emittedSuite.headerComment
  result += emittedSuite.dependencies
  result += `${generateNamespace(name)}`
  result += emittedSuite.suiteDeclaration
  result += emittedSuite.variables
  //result += emittedSuite.beforeAll
  result += emittedSuite.beforeEach
  result += emittedSuite.afterEach
  //result += emittedSuite.afterAll
  result += emittedSuite.methods
  if (emittedSuite.tests.testDeclaration) {
    const test = emittedSuite.tests
    result += test.testDeclaration
    result += test.inEachBegin
    result += test.commands
    result += test.inEachEnd
    result += test.testEnd
  } else {
    for (const testName in emittedSuite.tests) {
      const test = emittedSuite.tests[testName]
      result += test.testDeclaration
      result += test.inEachBegin
      result += test.commands
      result += test.inEachEnd
      result += test.testEnd
    }
  }
  result += emittedSuite.suiteEnd
  result += '}' // terminate namespace
  return result
}

export default language
