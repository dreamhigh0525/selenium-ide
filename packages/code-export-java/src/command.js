// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import exporter from '../../code-export-utils/src'
import location from './location'
import selection from './selection'

export const emitters = {
  addSelection: emitSelect,
  answerOnNextPrompt: skip,
  assert: emitAssert,
  assertAlert: emitAssertAlert,
  assertChecked: emitVerifyChecked,
  assertConfirmation: emitAssertAlert,
  assertEditable: emitVerifyEditable,
  assertElementPresent: emitVerifyElementPresent,
  assertElementNotPresent: emitVerifyElementNotPresent,
  assertNotChecked: emitVerifyNotChecked,
  assertNotEditable: emitVerifyNotEditable,
  assertNotSelectedValue: emitVerifyNotSelectedValue,
  assertNotText: emitVerifyNotText,
  assertPrompt: emitAssertAlert,
  assertSelectedLabel: emitVerifySelectedLabel,
  assertSelectedValue: emitVerifySelectedValue,
  assertValue: emitVerifyValue,
  assertText: emitVerifyText,
  assertTitle: emitVerifyTitle,
  check: emitCheck,
  chooseCancelOnNextConfirmation: skip,
  chooseCancelOnNextPrompt: skip,
  chooseOkOnNextConfirmation: skip,
  click: emitClick,
  clickAt: emitClick,
  close: emitClose,
  debugger: skip,
  //do: emitControlFlowDo,
  doubleClick: emitDoubleClick,
  doubleClickAt: emitDoubleClick,
  dragAndDropToObject: emitDragAndDrop,
  echo: emitEcho,
  editContent: emitEditContent,
  //else: emitControlFlowElse,
  //elseIf: emitControlFlowElseIf,
  //end: emitControlFlowEnd,
  executeScript: emitExecuteScript,
  executeAsyncScript: emitExecuteAsyncScript,
  //if: emitControlFlowIf,
  mouseDown: emitMouseDown,
  mouseDownAt: emitMouseDown,
  mouseMove: emitMouseMove,
  mouseMoveAt: emitMouseMove,
  mouseOver: emitMouseMove,
  mouseOut: emitMouseOut,
  mouseUp: emitMouseUp,
  mouseUpAt: emitMouseUp,
  open: emitOpen,
  //pause: emitPause,
  removeSelection: emitSelect,
  //repeatIf: emitControlFlowRepeatIf,
  //run: emitRun,
  //runScript: emitRunScript,
  select: emitSelect,
  //selectFrame: emitSelectFrame,
  //selectWindow: emitSelectWindow,
  sendKeys: emitSendKeys,
  //setSpeed: emitSetSpeed,
  setWindowSize: emitSetWindowSize,
  //store: emitStore,
  //storeAttribute: emitStoreAttribute,
  //storeText: emitStoreText,
  //storeTitle: emitStoreTitle,
  //storeValue: emitStoreValue,
  //storeWindowHandle: emitStoreWindowHandle,
  //storeXpathCount: emitStoreXpathCount,
  //submit: emitSubmit,
  //times: emitControlFlowTimes,
  type: emitType,
  uncheck: emitUncheck,
  verify: emitAssert,
  verifyChecked: emitVerifyChecked,
  verifyEditable: emitVerifyEditable,
  verifyElementPresent: emitVerifyElementPresent,
  verifyElementNotPresent: emitVerifyElementNotPresent,
  verifyNotChecked: emitVerifyNotChecked,
  verifyNotEditable: emitVerifyNotEditable,
  verifyNotSelectedValue: emitVerifyNotSelectedValue,
  verifyNotText: emitVerifyNotText,
  verifySelectedLabel: emitVerifySelectedLabel,
  verifySelectedValue: emitVerifySelectedValue,
  verifyText: emitVerifyText,
  verifyTitle: emitVerifyTitle,
  verifyValue: emitVerifyValue,
  //waitForElementEditable: emitWaitForElementEditable,
  //waitForElementPresent: emitWaitForElementPresent,
  //waitForElementVisible: emitWaitForElementVisible,
  //waitForElementNotEditable: emitWaitForElementNotEditable,
  //waitForElementNotPresent: emitWaitForElementNotPresent,
  //waitForElementNotVisible: emitWaitForElementNotVisible,
  //webdriverAnswerOnVisiblePrompt: emitAnswerOnNextPrompt,
  //webdriverChooseCancelOnVisibleConfirmation: emitChooseCancelOnNextConfirmation,
  //webdriverChooseCancelOnVisiblePrompt: emitChooseCancelOnNextConfirmation,
  //webdriverChooseOkOnVisibleConfirmation: emitChooseOkOnNextConfirmation,
  //while: emitControlFlowWhile,
}

exporter.preprocessors.register(emitters)

export function register(command, emitter) {
  emitters[command] = emitter
}

export function emit(command) {
  return exporter.command.emit(
    command,
    emitters[command.command],
    variableLookup
  )
}

function variableLookup(varName) {
  return `vars.get("${varName}").toString()`
}

function emitAssert(varName, value) {
  return Promise.resolve(
    `assertEquals(vars.get("${varName}").toString(), ${value});`
  )
}

function emitAssertAlert(alertText) {
  return Promise.resolve(`{
      Alert alert = driver.switchTo().alert();
      assertThat(alert.getText(), is("${alertText}"));
      alert.accept();
  }`)
}

async function emitCheck(locator) {
  return Promise.resolve(
    `{
        WebElement element = driver.findElement(${await location.emit(
          locator
        )});
        if (!element.isSelected()) {
          element.click();
        }
      }`
  )
}

async function emitClick(target) {
  return Promise.resolve(
    `driver.findElement(${await location.emit(target)}).click();`
  )
}

async function emitClose() {
  return Promise.resolve(`driver.close();`)
}

async function emitDoubleClick(target) {
  return Promise.resolve(
    `{
        WebElement element = driver.findElement(${await location.emit(target)});
        Actions builder = new Actions(driver);
        builder.doubleClick(element).perform();
      }`
  )
}

async function emitEcho(message) {
  return Promise.resolve(`System.out.println("${message}");`)
}

async function emitEditContent(locator, content) {
  return Promise.resolve(`
  {
      WebElement element = driver.findElement(${await location.emit(locator)});
      JavascriptExecutor js = (JavascriptExecutor) driver;
      js.executeScript("if(arguments[0].contentEditable === 'true') {arguments[0].innerHTML = '${content}'}", element);
  }`)
}

async function emitDragAndDrop(dragged, dropped) {
  return Promise.resolve(
    `{
        WebElement dragged = driver.findElement(${await location.emit(
          dragged
        )});
        WebElement dropped = driver.findElement(${await location.emit(
          dropped
        )});
        Actions builder = new Actions(driver);
        builder.dragAndDrop(dragged, dropped).perform();
      }`
  )
}

async function emitExecuteScript(script, varName) {
  return Promise.resolve(
    `{
      JavascriptExecutor js = (JavascriptExecutor) driver;
      ${varName ? `vars.push("${varName}") = ` : ''}js.executeScript("${
      script.script
    }${generateScriptArguments(script)}");
  }`
  )
}

async function emitExecuteAsyncScript(script, varName) {
  return Promise.resolve(
    `{
      JavascriptExecutor js = (JavascriptExecutor) driver;
      ${
        varName ? `vars.push("${varName}") =` : ''
      } js.executeAsyncScript("var callback = arguments[arguments.length - 1];${
      script.script
    }.then(callback).catch(callback);${generateScriptArguments(script)}");
    }`
  )
}

function generateScriptArguments(script) {
  return `${script.argv.length ? ',' : ''}${script.argv
    .map(varName => `vars.get("${varName}")`)
    .join(',')}`
}

async function emitMouseDown(locator) {
  return Promise.resolve(`
    {
        WebElement element = driver.findElement(${await location.emit(
          locator
        )});
        Action builder = new Actions(driver);
        builder.moveToElement(element).clickAndHold().perform();
    }`)
}

async function emitMouseMove(locator) {
  return Promise.resolve(`
    {
        WebElement element = driver.findElement(${await location.emit(
          locator
        )});
        Action builder = new Actions(driver);
        builder.moveToElement(element).perform();
    }`)
}

async function emitMouseOut() {
  return Promise.resolve(`
    {
        WebElement element = driver.findElement(By.tagName("body"));
        Action builder = new Actions(driver);
        builder.moveToElement(element, 0, 0).perform();
    }`)
}

async function emitMouseUp(locator) {
  return Promise.resolve(`
    {
        WebElement element = driver.findElement(${await location.emit(
          locator
        )});
        Action builder = new Actions(driver);
        builder.moveToElement(element).release().perform();
    }`)
}

function emitOpen(target) {
  const url = /^(file|http|https):\/\//.test(target)
    ? `"${target}"`
    : `"${global.baseUrl}${target}"`
  return Promise.resolve(`driver.get(${url});`)
}

async function emitSetWindowSize(size) {
  const [width, height] = size.split('x')
  return Promise.resolve(
    `driver.manage().window().setSize(new Dimension(${width}, ${height}));`
  )
}

async function emitType(target, value) {
  return Promise.resolve(
    `driver.findElement(${await location.emit(target)}).sendKeys("${value}");`
  )
}

async function emitSelect(selectElement, option) {
  return Promise.resolve(`{
    WebElement dropdown = driver.findElement(${await location.emit(
      selectElement
    )});
    dropdown.findElement(${await selection.emit(option)}).click();
  }`)
}

async function emitSendKeys(target, value) {
  return Promise.resolve(
    `driver.findElement(${await location.emit(target)}).sendKeys(${value
      .map(s => (s.startsWith('Key[') ? s : `"${s}"`))
      .join(',')}));`
  )
}

async function emitUncheck(locator) {
  return Promise.resolve(
    `{
        WebElement element = driver.findElement(${await location.emit(
          locator
        )});
        if (element.isSelected()) {
          element.click();
        }
      }`
  )
}

async function emitVerifyChecked(locator) {
  return Promise.resolve(
    `assertTrue(driver.findElement(${await location.emit(
      locator
    )}).isSelected());`
  )
}

async function emitVerifyEditable(locator) {
  return Promise.resolve(
    `{
        WebElement element = driver.findElement(${await location.emit(
          locator
        )});
        Boolean isEditable = element.isEnabled() && element.getAttribute("readonly") == null;
        assertTrue(isEditable)
    }`
  )
}

async function emitVerifyElementPresent(locator) {
  return Promise.resolve(`
  {
      List<WebElement> elements = driver.findElements(${await location.emit(
        locator
      )});
      assert(elements.size() > 0);
  }`)
}

async function emitVerifyElementNotPresent(locator) {
  return Promise.resolve(`
  {
      List<WebElement> elements = driver.findElements(${await location.emit(
        locator
      )});
      assert(elements.size() == 0);
  }`)
}

async function emitVerifyNotChecked(locator) {
  return Promise.resolve(
    `assertFalse(driver.findElement(${await location.emit(
      locator
    )}).isSelected());`
  )
}

async function emitVerifyNotEditable(locator) {
  return Promise.resolve(`
  {
      WebElement element = driver.findElement(${await location.emit(locator)});
      Boolean isEditable = element.isEnabled() && element.getAttribute("readonly") == null;
      assertFalse(isEditable)
  }`)
}

async function emitVerifyNotSelectedValue(locator, expectedValue) {
  return Promise.resolve(`
    {
        String value = driver.findElement(${await location.emit(
          locator
        )}).getAttribute("value");
        assertThat(value, is(not("${exporter.text.emit(expectedValue)}")));
    }`)
}

async function emitVerifyNotText(locator, text) {
  return Promise.resolve(`
  {
      String text = driver.findElement(${await location.emit(
        locator
      )}).getText();
      assertThat(text, is(not("${exporter.text.emit(text)}")));
  }`)
}

async function emitVerifySelectedLabel(locator, labelValue) {
  return Promise.resolve(`
    {
        WebElement element = driver.findElement(${await location.emit(
          locator
        )});
        String value = element.getAttribute("value");
        String locator = String.format("option[@value='%s']", value);
        String selectedText = element.findElement(By.xpath(locator)).getText();
        assertThat(selectedText, is("${labelValue}"));
    }`)
}

async function emitVerifySelectedValue(locator, value) {
  return emitVerifyValue(locator, value)
}

async function emitVerifyText(locator, text) {
  return Promise.resolve(
    `assertThat(driver.findElement(${await location.emit(
      locator
    )}).getText(), is("${exporter.text.emit(text)}"));`
  )
}

async function emitVerifyValue(locator, value) {
  return Promise.resolve(`
    {
        String value = driver.findElement(${await location.emit(
          locator
        )}).getAttribute("value");
        assertThat(value, is("${value}"));
    }`)
}

async function emitVerifyTitle(title) {
  return Promise.resolve(`assertThat(driver.getTitle(), is("${title}"));`)
}

function skip() {
  return Promise.resolve()
}

export default {
  emit,
}
