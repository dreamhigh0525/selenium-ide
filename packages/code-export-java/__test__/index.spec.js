// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import fs from 'fs'
import path from 'path'
import { emitTest, emitSuite, _emitMethod, _findTestByName } from '../src'
import { normalizeTestsInSuite } from '../../selenium-ide/src/neo/IO/normalize'

describe('Code Export Java Selenium', () => {
  it('should export a test to JUnit code', async () => {
    const project = JSON.parse(
      fs.readFileSync(path.join(__dirname, 'test-files', 'single-test.side'))
    )
    const results = await emitTest({
      baseUrl: project.url,
      test: project.tests[0],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a suite to JUnit code', async () => {
    const project = normalizeProject(
      JSON.parse(
        fs.readFileSync(path.join(__dirname, 'test-files', 'single-suite.side'))
      )
    )
    const results = await emitSuite({
      baseUrl: project.url,
      suite: project.suites[0],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a test to JUnit code with reused test method', async () => {
    const project = normalizeProject(
      JSON.parse(
        fs.readFileSync(
          path.join(__dirname, 'test-files', 'test-case-reuse.side')
        )
      )
    )
    const results = await emitTest({
      baseUrl: project.url,
      test: project.tests[1],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a suite to JUnit code with reused test method', async () => {
    const project = normalizeProject(
      JSON.parse(
        fs.readFileSync(
          path.join(__dirname, 'test-files', 'test-case-reuse.side')
        )
      )
    )
    const results = await emitSuite({
      baseUrl: project.url,
      suite: project.suites[0],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
})

function normalizeProject(project) {
  let _project = { ...project }
  _project.suites.forEach(suite => {
    normalizeTestsInSuite({ suite, tests: _project.tests })
  })
  return _project
}
