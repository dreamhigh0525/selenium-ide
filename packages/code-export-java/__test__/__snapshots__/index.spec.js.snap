// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code Export Java Selenium should export a suite to JUnit code 1`] = `
"
import org.junit.Test;
import org.junit.Before;
import org.junit.After;
import static org.junit.Assert.*;
import static org.hamcrest.CoreMatchers.is;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.Dimension;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import java.util.HashMap;
import java.util.List;

public class Login {
    private WebDriver driver;
    private HashMap<String, Object> vars = new HashMap<>();

    @BeforeAll
    public void initialSetUp() {
    }

    @Before
    public void setUp() {
        driver = new FirefoxDriver();
    }

    @After
    public void tearDown() {
        driver.quit();
    }

    @AfterAll
    public void finalTearDown() {
    }

    @Test
    public void validcredentials() {
        driver.get(\\"http://the-internet.herokuapp.com/login\\");
        driver.manage().window().setSize(new Dimension(1440, 1177));
        driver.findElement(By.id(\\"username\\")).sendKeys(\\"tomsmith\\");
        driver.findElement(By.id(\\"password\\")).sendKeys(\\"SuperSecretPassword!\\");
        driver.findElement(By.cssSelector(\\"#login button\\")).click();
        assertThat(driver.findElement(By.cssSelector(\\".flash.success\\")).getText(), is(\\"You logged into a secure area!\\\\\\\\n×\\"));
    }

    @Test
    public void invalidcredentials() {
        driver.get(\\"http://the-internet.herokuapp.com/login\\");
        driver.manage().window().setSize(new Dimension(1440, 1177));
        driver.findElement(By.id(\\"username\\")).sendKeys(\\"blah\\");
        driver.findElement(By.id(\\"password\\")).sendKeys(\\"blah\\");
        driver.findElement(By.cssSelector(\\"#login button\\")).click();
        assertThat(driver.findElement(By.id(\\"flash\\")).getText(), is(\\"Your username is invalid!\\\\\\\\n×\\"));
    }
}
"
`;

exports[`Code Export Java Selenium should export a test to JUnit code 1`] = `
"
import org.junit.Test;
import org.junit.Before;
import org.junit.After;
import static org.junit.Assert.*;
import static org.hamcrest.CoreMatchers.is;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.Dimension;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import java.util.HashMap;
import java.util.List;

public class Login {
    private WebDriver driver;
    private HashMap<String, Object> vars = new HashMap<>();

    @BeforeAll
    public void initialSetUp() {
    }

    @Before
    public void setUp() {
        driver = new FirefoxDriver();
    }

    @After
    public void tearDown() {
        driver.quit();
    }

    @AfterAll
    public void finalTearDown() {
    }

    @Test
    public void login() {
        driver.get(\\"http://the-internet.herokuapp.com/login\\");
        driver.manage().window().setSize(new Dimension(1440, 1177));
        driver.findElement(By.id(\\"username\\")).sendKeys(\\"tomsmith\\");
        driver.findElement(By.id(\\"password\\")).sendKeys(\\"SuperSecretPassword!\\");
        driver.findElement(By.cssSelector(\\"#login button\\")).click();
        assertThat(driver.findElement(By.cssSelector(\\".flash.success\\")).getText(), is(\\"You logged into a secure area!\\\\\\\\n×\\"));
    }
}
"
`;
